stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,654 @@
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import json
+import os
+import shutil
+import tempfile
+from abc import abstractmethod
+
+import boto3
+from botocore.exceptions import ClientError
+
+from stream_alert.shared.helpers.boto import default_config
+from stream_alert.shared.logger import get_logger
+from stream_alert_cli.outputs.helpers import kms_encrypt, send_creds_to_s3
+
+LOGGER = get_logger(__name__)
+
+
+class OutputCredentialsProvider(object):
+    """Loads credentials that are housed on AWS S3, or cached locally.
+
+    Helper service to OutputDispatcher.
+
+    OutputDispatcher implementations may require credentials to authenticate with an external
+    gateway. All credentials for OutputDispatchers are to be stored in a single bucket on AWS S3
+    and are encrypted with AWS KMS. When alerts are dispatched via OutputDispatchers, these
+    encrypted credentials are downloaded and cached locally on the filesystem. Then, AWS KMS is
+    used to decrypt the credentials when in use.
+
+    Public methods:
+        load_credentials: Returns a dict of the credentials requested
+        get_local_credentials_temp_dir(): Returns full path to a temporary directory where all
+            encrypted credentials are cached.
+
+    """
+
+    def __init__(self, config, defaults, region, service_name, prefix=None, aws_account_id=None):
+        self._config = config
+        self._region = region
+        self._defaults = defaults
+        self._service_name = service_name
+
+        # Dependency on os package
+        if prefix is None:
+            prefix = os.environ['STREAMALERT_PREFIX']
+
+        if aws_account_id is None:
+            aws_account_id = os.environ['AWS_ACCOUNT_ID']
+
+        self._prefix = prefix
+        self._account_id = aws_account_id
+
+        # Drivers are strategies utilized by this class for fetching credentials from various
+        # locations on disk or remotely
+        self._drivers = []  # type: list[CredentialsProvidingDriver]
+        self._core_driver = None  # type: S3Driver
+        self._setup_drivers()
+
+    def _setup_drivers(self):
+        """Initializes all drivers.
+
+        The Drivers are sequentially checked in the order they are appended to the driver list.
+        """
+
+        # Ephemeral driver
+        ep_driver = EphemeralUnencryptedDriver(self._service_name)
+        self._drivers.append(ep_driver)
+
+        # Fall back onto downloading encrypted credentials from S3
+        s3_driver = S3Driver(self._prefix, self._service_name, self._region, cache_driver=ep_driver)
+        self._core_driver = s3_driver
```

Comment:
it seems like the `core_driver` is designed for 'remote' interaction.. ie, reading/writing from S3. might it be more appropriate to name something like `remote_driver` or similar? thoughts?

Reply:
I thought of it like a "Driver who stores the source of Truth" kind of thing. I was thinking this would be where the SSMDriver would replace the S3 driver as the "primary" one. I'd think that different Outputs would have different Core drivers depending on where the credentials are configured (if we do a slow migration).